source/plugins/ruby/arc_k8s_cluster_identity.rb

```diff
@@ -0,0 +1,210 @@
+        status = JSON.parse(get_response.body)["status"]
+        tokenReference["expirationTime"] = status["expirationTime"]
+        tokenReference["secretName"] = status["tokenReference"]["secretName"]
+        tokenReference["dataName"] = status["tokenReference"]["dataName"]
+      end
+    rescue => err
+      @log.warn "get_token_reference_from_crd call failed: #{err}"
+    end
+    return tokenReference
+  end
+
+  private
+
+  def renew_near_expiry_token()
+    begin
+      crd_request_uri = @@crd_resource_uri_template % {
+        kube_api_server_url: @kube_api_server_url,
+        cluster_config_crd_api_version: @@cluster_config_crd_api_version,
+        cluster_identity_resource_namespace: @@cluster_identity_resource_namespace,
+        cluster_identity_resource_name: @@cluster_identity_resource_name,
+      }
+      crd_request_body = get_crd_request_body
+      crd_request_body_json = crd_request_body.to_json
+      update_request = Net::HTTP::Patch.new(crd_request_uri)
+      update_request["Content-Type"] = "application/merge-patch+json"
+      update_request["Authorization"] = "Bearer #{@service_account_token}"
+      update_request.body = crd_request_body_json
+      update_response = @http_client.request(update_request)
+      @log.info "Got response of #{update_response.code} for PATCH #{crd_request_uri} @ #{Time.now.utc.iso8601}"
+      if update_response.code.to_i == 404
+        @log.info "since crd resource doesnt exist since creating crd resource : #{@@cluster_identity_resource_name} @ #{Time.now.utc.iso8601}"
+        create_request = Net::HTTP::Post.new(crd_request_uri)
+        create_request["Content-Type"] = "application/json"
+        create_request["Authorization"] = "Bearer #{@service_account_token}"
+        create_request.body = crd_request_body_json
+        create_response = @http_client.request(create_request)
+        @log.info "Got response of #{create_response.code} for POST #{crd_request_uri} @ #{Time.now.utc.iso8601}"
+      end
+    rescue => err
+      @log.warn "renew_near_expiry_token call failed: #{err}"
+    end
+  end
+
+  private
+
+  def get_service_account_token()
+    begin
+      if File.exist?(@token_file_path) && File.readable?(@token_file_path)
+        token_str = File.read(@token_file_path).strip
+        return token_str
+      else
+        @log.warn "Unable to read token string from #{@token_file_path}"
+        return nil
+      end
+    end
+  end
+
+  private
+
+  def get_http_client()
+    kube_api_server_url = get_kube_api_server_url
+    base_api_server_url = URI.parse(kube_api_server_url)
+    http = Net::HTTP.new(base_api_server_url.host, base_api_server_url.port)
+    http.use_ssl = true
+    if !File.exist?(@cert_file_path)
+      raise "#{@cert_file_path} doesnt exist"
+    else
+      http.ca_file = @cert_file_path
+    end
+    http.verify_mode = OpenSSL::SSL::VERIFY_PEER
+    return http
+  end
+
+  private
+
+  def get_kube_api_server_url
```

Comment:
pls move this to KubernetesApiClient

Reply:
> pls move this to KubernetesApiClient

Moved this code KubernetesApiClient and added begin & rescue block though no chance of getting this fail.